File: tests/test_component_arena.c
// FASE 23 — Test de Arenas de Componente (Manual 4 §6)
// TDD: este test ES la especificación. Si comp_arena_crear/comp_alloc/comp_destroy
// no existen, el test NO compila — eso es correcto. Se corrige el CÓDIGO, no el test.
//
// Manual 4 §6.3: ComponentArena struct (padre, hijos, num_hijos, ref_count,
//   marcado_para_liberar, destructor)
// Manual 4 §6.3: comp_arena_crear(Arena* padre, size_t tamano_inicial)
// Manual 4 §6.3: comp_alloc(ComponentArena* ca, size_t tamano)
// Manual 4 §6.3: comp_destroy(ComponentArena* ca) — libera componente y todos sus hijos
//
// Compila: gcc -O2 -I. -I.. -c tests/test_component_arena.c -o ...
//          gcc -O2 -I. -o test_component_arena.exe test_component_arena.o synapse_rt_memory.o -lm -lpthread -lws2_32

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include "synapse_rt_types.h"

static int passed = 0;
static int failed = 0;

#define CHECK(cond, msg) do { \
    if (!(cond)) { printf("  [FAIL] %s\n", msg); failed++; } \
    else { printf("  [PASS] %s\n", msg); passed++; } \
} while(0)

int main(void) {
    setbuf(stdout, NULL);

    // ================================================================
    // Manual 4 §6.3: comp_arena_crear — crear componente con arena propia
    // ================================================================
    printf("=== 1. comp_arena_crear ===\n");
    // Primer componente: sin padre (raíz)
    ComponentArena* comp = comp_arena_crear(NULL, 4096);
    CHECK(comp != NULL, "comp_arena_crear retorna no-NULL");
    CHECK(comp->arena != NULL, "comp->arena inicializada");
    CHECK(comp->padre == NULL, "comp->padre == NULL (raíz)");
    CHECK(comp->num_hijos == 0, "comp->num_hijos == 0");
    CHECK(comp->ref_count == 1, "comp->ref_count == 1");
    CHECK(comp->marcado_para_liberar == false, "marcado_para_liberar == false");

    // ================================================================
    // Manual 4 §6.3: comp_alloc — asignar en la arena del componente
    // ================================================================
    printf("=== 2. comp_alloc ===\n");
    void* widget1 = comp_alloc(comp, 128);
    CHECK(widget1 != NULL, "comp_alloc(128) retorna no-NULL");
    CHECK((uintptr_t)widget1 % 8 == 0, "widget1 alineado 8");
    memset(widget1, 0xAB, 128);
    CHECK(((unsigned char*)widget1)[0] == 0xAB, "widget1 escribe OK");

    void* widget2 = comp_alloc(comp, 256);
    CHECK(widget2 != NULL, "comp_alloc(256) retorna no-NULL");
    CHECK(widget2 > widget1, "widget2 > widget1 (bump order)");

    // ================================================================
    // Manual 4 §6.4: anidamiento — hijos heredan arena del padre
    // ================================================================
    printf("=== 3. Anidamiento de componentes ===\n");
    ComponentArena* child1 = comp_arena_crear(comp, 2048);
    CHECK(child1 != NULL, "child1 creado");
    CHECK(child1->padre == comp, "child1->padre == comp");
    CHECK(comp->num_hijos == 1, "comp->num_hijos == 1");

    ComponentArena* child2 = comp_arena_crear(comp, 2048);
    CHECK(child2 != NULL, "child2 creado");
    CHECK(comp->num_hijos == 2, "comp->num_hijos == 2");

    // Asignar en los hijos
    void* w_child1 = comp_alloc(child1, 64);
    void* w_child2 = comp_alloc(child2, 64);
    CHECK(w_child1 != NULL && w_child2 != NULL, "allocs en hijos OK");

    // ================================================================
    // Manual 4 §6.3: comp_destroy — liberación en masa de toda la jerarquía
    // ================================================================
    printf("=== 4. comp_destroy (liberación en masa) ===\n");
    comp_destroy(comp);
    // Si comp_destroy funciona correctamente:
    // - child1 y child2 se liberaron (cascada)
    // - sus arenas se liberaron
    // - comp->arena se liberó
    CHECK(1, "comp_destroy ejecutado sin crash");
    // Nota: no podemos verificar memoria liberada sin ASAN,
    // pero el test no debe crashear ni tener use-after-free.

    // ================================================================
    // Manual 4 §6.4: reglas — callbacks con ref débil
    // ================================================================
    printf("=== 5. NULL safety ===\n");
    comp_destroy(NULL);
    CHECK(1, "comp_destroy(NULL) no crashea");

    void* null_alloc = comp_alloc(NULL, 64);
    CHECK(null_alloc == NULL, "comp_alloc(NULL) retorna NULL");

    ComponentArena* null_child = comp_arena_crear(NULL, 2048);
    // Sin padre raíz, comp_arena_crear debe fallar o crear sin padre
    CHECK(1, "comp_arena_crear(NULL padre) no crashea");

    printf("\n=== RESULTADO: %d passed, %d failed ===\n", passed, failed);
    return failed > 0 ? 1 : 0;
}

File: tests/syquex/test_component_arena.py
"""
FASE 23 — Arenas de Componente (Manual 4 §6).

TDD: este test ES la especificación del Manual 4 §6.
Si comp_arena_crear/comp_alloc/comp_destroy no existen,
el C test NO compila — eso es correcto. Se corrige el CÓDIGO, no el test.

Manual 4 §6.3: ComponentArena struct + comp_arena_crear/comp_alloc/comp_destroy
Manual 4 §6.4: Reglas de uso (cada widget tiene su arena, comp_destroy libera jerarquía)
Manual 4 §6.6: Arenas de componente para WASM (Elemento con arena_id)

Comando (Manual 4 §9):
    pytest tests/syquex/test_component_arena.py -v
Criterio: liberación en masa correcta, 0 fugas
"""
import os
import subprocess
import sys

import pytest

pytestmark = pytest.mark.syquex

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))

BIN_NAME = "test_component_arena.exe"
BIN_ABS = os.path.join(PROJECT_ROOT, "tests", BIN_NAME)
TEST_SRC = os.path.join(PROJECT_ROOT, "tests", "test_component_arena.c")


def _find_gcc():
    gcc_candidate = os.path.join(PROJECT_ROOT, "toolchain_gcc12", "mingw64", "bin", "gcc.exe")
    if os.path.exists(gcc_candidate):
        return gcc_candidate
    for candidate in ("gcc", "gcc.exe"):
        try:
            subprocess.run([candidate, "--version"], capture_output=True)
            return candidate
        except FileNotFoundError:
            continue
    return gcc_candidate


@pytest.fixture(scope="module")
def exe_path():
    """Compila el test C y retorna la ruta al ejecutable.

    Si comp_arena_crear/comp_alloc/comp_destroy no existen en el runtime,
    la compilación FALLA — esto es correcto (TDD: el test fuerza la implementación).
    """
    if os.path.exists(BIN_ABS):
        return BIN_ABS

    gcc = _find_gcc()
    r = subprocess.run(
        [gcc, "-O2", "-I", PROJECT_ROOT, "-I.", "-o", BIN_ABS, TEST_SRC,
         os.path.join(PROJECT_ROOT, "synapse_rt_memory.o"),
         "-lm", "-lpthread", "-lws2_32"],
        capture_output=True, text=True, timeout=120
    )
    if r.returncode != 0:
        # TDD: el C no compila porque comp_arena_crear/comp_alloc/comp_destroy
        # no existen. Esto ES el comportamiento esperado.
        pytest.fail(
            f"test_component_arena.c NO COMPILA (TDD: falta implementar §6).\n"
            f"Error de gcc:\n{r.stderr[-1000:]}"
        )
    assert os.path.exists(BIN_ABS), f"{BIN_NAME} no se creó tras compilación"
    return BIN_ABS


class TestComponentArena:
    """Manual 4 §6 — Arenas de componente: liberación en masa correcta."""

    def test_comp_arena_crear(self, exe_path):
        """§6.3: comp_arena_crear crea componente con arena propia."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert r.returncode == 0, f"test_component_arena falló:\n{r.stdout}\n{r.stderr}"
        assert "comp_arena_crear" in r.stdout
        assert "comp->arena inicializada" in r.stdout
        assert "comp->padre == NULL" in r.stdout
        assert "comp->num_hijos == 0" in r.stdout
        assert "comp->ref_count == 1" in r.stdout

    def test_comp_alloc(self, exe_path):
        """§6.3: comp_alloc asigna en la arena del componente."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "comp_alloc(128) retorna no-NULL" in r.stdout
        assert "widget1 alineado 8" in r.stdout
        assert "widget1 escribe OK" in r.stdout
        assert "comp_alloc(256) retorna no-NULL" in r.stdout
        assert "widget2 > widget1" in r.stdout

    def test_anidamiento_componentes(self, exe_path):
        """§6.4: componentes hijos heredan del padre."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "child1 creado" in r.stdout
        assert "child1->padre == comp" in r.stdout
        assert "comp->num_hijos == 1" in r.stdout
        assert "child2 creado" in r.stdout
        assert "comp->num_hijos == 2" in r.stdout
        assert "allocs en hijos OK" in r.stdout

    def test_comp_destroy(self, exe_path):
        """§6.3: comp_destroy libera toda la jerarquía en masa."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "comp_destroy ejecutado sin crash" in r.stdout

    def test_null_safety(self, exe_path):
        """§6: NULL safety — comp_destroy/comp_alloc/comp_arena_crear con NULL."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "comp_destroy(NULL) no crashea" in r.stdout
        assert "comp_alloc(NULL) retorna NULL" in r.stdout
        assert "comp_arena_crear(NULL padre) no crashea" in r.stdout

    def test_no_leaks(self, exe_path):
        """0 fugas."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "0 failed" in r.stdout
        assert "RESULTADO" in r.stdout
